ego/cli/elf_test.go

```diff
@@ -0,0 +1,200 @@
+	assert := assert.New(t)
+
+	// Setup test environment
+	fs := afero.Afero{Fs: afero.NewMemMapFs()}
+	runner := signRunner{fs: fs}
+	cli := NewCli(&runner, fs)
+
+	// Copy from hostfs to memfs
+	const exe = "helloworld"
+	require.NoError(fs.WriteFile(exe, elfUnsigned, 0))
+
+	// Check if no payload exists
+	unsignedExeMemfs, err := fs.Open(exe)
+	require.NoError(err)
+	unsignedExeMemfsStat, err := unsignedExeMemfs.Stat()
+	require.NoError(err)
+	unsignedExeMemfsSize := unsignedExeMemfsStat.Size()
+	payloadSize, payloadOffset, oeInfoOffset, err := getPayloadInformation(unsignedExeMemfs)
+	assert.NoError(err)
+	assert.Zero(payloadSize)
+	assert.Zero(payloadOffset)
+	assert.NotZero(oeInfoOffset)
+	unsignedExeMemfs.Close()
+
+	// Create a default config we want to check
+	testConf := &config.Config{
+		Exe:             exe,
+		Key:             defaultPrivKeyFilename,
+		Debug:           true,
+		HeapSize:        512, //[MB]
+		ProductID:       1,
+		SecurityVersion: 1,
+	}
+
+	// Marshal config
+	jsonData, err := json.Marshal(testConf)
+	require.NoError(err)
+	expectedLengthOfPayload := len(jsonData)
+
+	// Embed json config to helloworld
+	err = cli.embedConfigAsPayload(exe, jsonData)
+	assert.NoError(err)
+
+	// Check if new helloworld_signed now contains signed data
+	signedExeMemfs, err := fs.Open(exe)
+	require.NoError(err)
+	defer signedExeMemfs.Close()
+
+	payloadSize, payloadOffset, oeInfoOffset, err = getPayloadInformation(signedExeMemfs)
+	assert.NoError(err)
+	assert.EqualValues(expectedLengthOfPayload, payloadSize)
+	assert.EqualValues(unsignedExeMemfsSize, payloadOffset)
+	assert.NotZero(oeInfoOffset)
+
+	// Reconstruct the JSON and check equality
+	reconstructedJSON := make([]byte, payloadSize)
+	n, err := signedExeMemfs.ReadAt(reconstructedJSON, payloadOffset)
+	require.NoError(err)
+	require.EqualValues(payloadSize, n)
+	assert.EqualValues(jsonData, reconstructedJSON)
+
+	// Now modify the config, redo everything and see if trucate worked fine and everything still lines up
+	testConf.HeapSize = 5120
+	jsonNewData, err := json.Marshal(testConf)
+	require.NoError(err)
+	expectedLengthOfNewPayload := len(jsonNewData)
+
+	// Re-sign the already signed executable
+	err = cli.embedConfigAsPayload(exe, jsonNewData)
+	assert.NoError(err)
+	payloadSize, payloadOffset, _, err = getPayloadInformation(signedExeMemfs)
+	require.NoError(err)
+	assert.EqualValues(expectedLengthOfNewPayload, payloadSize)
+	assert.EqualValues(unsignedExeMemfsSize, payloadOffset)
+
+	// Reconstruct the JSON and check if it not the old one, but the new one
+	reconstructedJSON = make([]byte, payloadSize)
+	n, err = signedExeMemfs.ReadAt(reconstructedJSON, payloadOffset)
+	require.NoError(err)
+	require.EqualValues(payloadSize, n)
+
+	// Finally, check if we got the new JSON config and not the old one
+	assert.NotEqualValues(jsonData, reconstructedJSON)
+	assert.EqualValues(jsonNewData, reconstructedJSON)
+}
+
+func TestCheckUnsupportedImports(t *testing.T) {
```

Comment:
Yes, it's quite heavy and also has some prerequisites on the system. Can we instead append a few lines to src/integration_test.sh to cover this feature?